src/main/java/org/skriptlang/skript/lang/condition/DNFConditionalBuilder.java

```diff
@@ -0,0 +1,172 @@
+package org.skriptlang.skript.lang.condition;
+
+import com.google.common.base.Preconditions;
+import org.jetbrains.annotations.Contract;
+import org.skriptlang.skript.lang.condition.Conditional.Operator;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static org.skriptlang.skript.lang.condition.Conditional.negate;
+
+/**
+ * Builds a Disjunctive Normal Form {@link CompoundConditional}, meaning it is solely composed of groups of ANDs all ORed together,
+ * ex. {@code (a && !b && c) || b || (!c && d)}.
+ * <br>
+ * A builder should no longer be used after calling {@link #build()}.
+ * @param <T> The context class to use for evaluation.
+ */
+public class DNFConditionalBuilder<T> {
+
+	private CompoundConditional<T> root;
+
+	/**
+	 * Creates an empty builder.
+	 */
+	DNFConditionalBuilder() {
+		this.root = null;
+	}
+
+	/**
+	 * Creates a builder with a single conditional.
+	 * @param root If given a {@link CompoundConditional}, it is used as the root conditional.
+	 *             Otherwise, a OR compound conditional is created as the root with the given conditional within.
+	 */
+	DNFConditionalBuilder(Conditional<T> root) {
+		if (root instanceof CompoundConditional<T> compoundConditional) {
+			this.root = compoundConditional;
+		} else {
+			this.root = new CompoundConditional<>(Operator.OR, root);
+		}
+	}
+
+	/**
+	 * @return The root conditional, which will be DNF-compliant
+	 * @throws IllegalStateException if the builder is empty.
+	 */
+	public Conditional<T> build() {
+		Preconditions.checkState(root != null, "Cannot build an empty conditional!");
+		return root;
```

Comment:
We can simplify this further 🙂 
```suggestion
		return Preconditions.checkNotNull(root, "Cannot build an empty conditional!");
```